lib/Storage/S3StorageClient.cpp

```diff
@@ -0,0 +1,423 @@
+    return std::make_shared<LoggerSystem>(Level);
+  };
+  Aws::InitAPI(Options);
+  OnQuit->add([Options = std::move(Options)] { Aws::ShutdownAPI(Options); });
+}
+
+} // namespace
+
+static llvm::StringRef consumeSplit(llvm::StringRef &Input, char SplitChar) {
+  std::size_t Pos = Input.find(SplitChar);
+  revng_assert(Pos != llvm::StringRef::npos);
+  llvm::StringRef Result = Input.slice(0, Pos);
+  Input = Input.drop_front(Pos + 1);
+  return Result;
+}
+
+template<typename T>
+static llvm::Error
+toError(const Aws::Utils::Outcome<T, Aws::S3::S3Error> &Request) {
+  revng_check(!Request.IsSuccess());
+  return llvm::createStringError(llvm::inconvertibleErrorCode(),
+                                 Request.GetError().GetMessage());
+}
+
+static std::string resolvePath(const llvm::StringRef Root,
+                               const llvm::StringRef Path) {
+  if (Root.empty())
+    return Path.str();
+  else if (Path.empty())
+    return Root.str();
+  else
+    return joinPath(Root, Path);
+}
+
+static std::string generateNewFilename(const llvm::StringRef Path) {
+  revng_assert(not Path.ends_with("/"));
+
+  llvm::StringRef Base;
+  llvm::StringRef Filename;
+  if (Path.contains("/")) {
+    auto Parts = Path.rsplit("/");
+    Base = Parts.first;
+    Filename = Parts.second;
+  } else {
+    Filename = Path;
+  }
+
+  using uuid_t = uint8_t[16];
+  uuid_t UUID;
+  for (int I = 0; I < 16; I++)
+    UUID[I] = (llvm::sys::Process::GetRandomNumber() % UINT8_MAX);
+
+  std::string Output;
+  llvm::raw_string_ostream OS(Output);
+  OS.write_uuid(UUID);
+  OS.flush();
+
+  return Base.str() + "/" + Output + "-" + Filename.str();
+}
+
+class S3ReadableFile : public ReadableFile {
+private:
+  TemporaryFile TempFile;
+  std::unique_ptr<llvm::MemoryBuffer> Buffer;
+
+public:
+  S3ReadableFile(TemporaryFile &&TempFile,
+                 std::unique_ptr<llvm::MemoryBuffer> &&Buffer) :
+    TempFile(std::move(TempFile)), Buffer(std::move(Buffer)) {}
+  ~S3ReadableFile() override = default;
+  llvm::MemoryBuffer &buffer() override { return *Buffer; };
+};
+
+class S3WritableFile : public WritableFile {
+private:
+  TemporaryFile TempFile;
+  std::unique_ptr<llvm::raw_fd_ostream> OS;
+  std::string Path;
+  ContentEncoding Encoding;
+  S3StorageClient &Client;
+
+public:
+  S3WritableFile(TemporaryFile &&TempFile,
+                 std::unique_ptr<llvm::raw_fd_ostream> &&OS,
+                 llvm::StringRef Path,
+                 ContentEncoding Encoding,
+                 S3StorageClient &Client) :
+    TempFile(std::move(TempFile)),
+    OS(std::move(OS)),
+    Path(Path.str()),
+    Encoding(Encoding),
+    Client(Client) {}
+
+  llvm::raw_pwrite_stream &os() override { return *OS; }
+  llvm::Error commit() override {
+    OS->flush();
+
+    Aws::S3::Model::PutObjectRequest Request;
+    Request.SetBucket(Client.Bucket);
+
+    if (Encoding == ContentEncoding::Gzip)
+      Request.SetContentEncoding("gzip");
```

Comment:
This only has the effect of letting the S3 server known that has to serve this file with the proper content-encoding, correct? Maybe add a comment.

Reply:
> correct?

Yes

> Maybe add a comment

Will do

Reply:
Added comments